worker/src/ws.rs

```diff
@@ -0,0 +1,175 @@
+                    if let Some(text) = msg.text() {
+                        if let Err(e) = dest_ws.send_with_str(text) {
+                            close_with_internal_error(format!("Error sending to destination with string: {:?}", e), &dest_ws, &source_ws_server);
+                            break;
+                        }
+                    } else if let Some(bytes) = msg.bytes() {
+                        if let Err(e) = dest_ws.send_with_bytes(bytes) {
+                            close_with_internal_error(format!("Error sending to destination with bytes: {:?}", e), &dest_ws, &source_ws_server);
+                            break;
+                        }
+                    } else {
+                        close_with_internal_error(format!("Error message from source with no text or bytes"), &dest_ws, &source_ws_server);
+                        break;
+                    }
+                }
+                Ok(WebsocketEvent::Close(close)) => {
+                    console_log!("Close event from source: {:?}", close);
+                    let _ = dest_ws.close(Some(close.code()), Some(close.reason()));
+                    break;
+                }
+                Err(e) => {
+                    close_with_internal_error(format!("Other source websocket error: {:?}", e), &dest_ws, &source_ws_server);
+                    break;
+                }
+            },
+            Either::Right((Some(dest_event), _)) => match dest_event {
+                Ok(WebsocketEvent::Message(msg)) => {
+                    if let Some(text) = msg.text() {
+                        if let Err(e) = source_ws_server.send_with_str(text) {
+                            close_with_internal_error(format!("Error sending to source with string: {:?}", e), &dest_ws, &source_ws_server);
+                            break;
+                        }
+                    } else if let Some(bytes) = msg.bytes() {
+                        if let Err(e) = source_ws_server.send_with_bytes(bytes) {
+                            close_with_internal_error(format!("Error sending to source with bytes: {:?}", e), &dest_ws, &source_ws_server);
+                            break;
+                        }
+                    } else {
+                        close_with_internal_error(format!("Error message from destination with no text or bytes"), &dest_ws, &source_ws_server);
+                        break;
+                    }
+                }
+                Ok(WebsocketEvent::Close(close)) => {
+                    console_log!("Close event from destination: {:?}", close);
+                    let _ = source_ws_server.close(Some(close.code()), Some(close.reason()));
+                    break;
+                }
+                Err(e) => {
+                    close_with_internal_error(format!("Other destination websocket error: {:?}", e), &dest_ws, &source_ws_server);
+                    break;
+                }
```

Comment:
This big ol thing is a mess. Both blocks do more or less the same thing and implement the same error logic.. I can probably pull it out into a function..